docs/combinatorics.rst

```diff
@@ -1,6 +1,202 @@
+    the same but whose children are ordered differently.
+
+However, this method has the potential to double-count trees when we encounter
+partitions with repeated numbers of leaves. To remove these redundant trees we
+reduce the set of trees so that if a subtree has the same number of leaves as
+its left sibling, it must have an equal or greater shape rank.
+
+In the case of binary trees, this simplifies to:
+Let :math:`P = [j, k]` be a partition of :math:`n`. If :math:`j < k`,
+then the trees generated by :math:`P` are isomorphic to :math:`S_j \times S_k`.
+If :math:`j = k`, then the trees generated are isomorphic to all 2-element
+combinations with replacement of trees from :math:`S_j`.
+This generalizes to the following algorithm run per partition:
+
+1.
+    Group each subsequence of equal elements in a partition :math:`P` of :math:`n`.
+2. 
+    For a group :math:`G` whose elements equal :math:`k`, compute all
+    combinations with replacement of :math:`|G|` elements from :math:`S_k`
+3.
+    Take the cartesian product of all groups
+
+.. note::
+    By grouping the partition of :math:`n` into symmetrical groups, we eliminate
+    symmetry between groups, allowing us to take the cartesian product between
+    groups of trees. This is a pattern that emerges again in labelling.
+
+*********************
+Labelling Tree Shapes
+*********************
+
+Tree shapes are useful in and of themselves, but we can use the enumeration
+formulated above to go further and enumerate all trees with labelled leaves.
+We now have the challenge of, given a tree shape, generating all unique
+labellings of that tree.
+We cannot have two distinct labellings where, under some mirroring of subtrees,
+two trees become identical.
+
+Let :math:`T` be a tree in :math:`S_n` generated by the partition
+:math:`P` of :math:`n`.
+Here, we face a similar problem of symmetry as in the unlabelled
+case. In this case, the problem arises when :math:`T` has identical subtrees,
+as switching those two trees would not violate the canonical
+orientation of the tree but would result in two different
+labellings.
+
+To remove this redundancy, we enforce that if two subtrees are equivalent,
+the left tree must have the minimum label out of all labels across the two
+subtrees. Below are the different ways to label a single tree shape
+in :math:`S_3`.
+
+.. image:: _static/tree_labelling_0.svg
+    :width: 32%
+.. image:: _static/tree_labelling_1.svg
+    :width: 32%
+.. image:: _static/tree_labelling_2.svg
+    :width: 32%
+
+Note that the right subtree is symmetrical, so we must always give the smaller
+label to the left leaf in that subtree to avoid redundant labellings.
+
+This *canonical labelling* removes any redundant trees and allows us the
+following enumeration algorithm, which is presented below in both the
+binary and general case.
+
+Examining first the binary case, let :math:`T = (T_j, T_k)`, with left subtree
```

Comment:
At this point, I'm not sure that the algorithm is what we want. Probably best to keep this level of detail for a more technical writeup. I guess what we're trying to get across here in these docs is, *what* it does, not *how* it does it.

Reply:
Agree - I don't think this is an enumeration algorithm? I see what the ordering *is*, but not everything you say below comes in a natural order, as far as I can tell.

Oh, wait: actually, the ordering itself is a bit confusing, as it's defined implicitly in terms of the orderings of smaller-n labelings? You've said how to compare labelings with different labels; but to rank labelings you need to also compare labellings with the same labels. Maybe inserting a definition along the lines of the one above would be good. I guess that this is that the rank of an labeling of a tree is its rank among sorted tuples `((size_1, shape_1, labeling_rank_1), ... (size_1, shape_1, labeling_rank_1))`, where `labeling_rank_1` is the rank of that particular labeling of labels [1...n] to the subtree with shape `shape_1` and size `size_1` (which is less than `n`)?

I may not have been very clear there. I am guessing that the labeling
```
  ┃  
 ┏┻━┓
 ┃  ┃
┏┻┓ ┃
┃ ┃ ┃
1 2 0
```
comes before
```
  ┃  
 ┏┻━┓
 ┃  ┃
┏┻┓ ┃
┃ ┃ ┃
0 2 1
```
because the smallest (and thus earliest) subtree has the smaller label in the first labeling?  Is this true?

Reply:
@petrelharp I *think* I understand what you're asking and yes, there is a piece missing here. The way in which labellings are created is directly dependent on how you choose combinations of labels when splitting labels up between subtrees. This uses a lexicographic ordering for choose that comes with `itertools.combinations` to select labels to assign to subtrees, so for the trees you mention it would be ordered (0, (1, 2)), (1, (0, 2)), (2, (0, 1))

Reply:
Ah-ha - that's equivalent to what I said, but simpler. So, you're saying that labelings are ordered lexicographically, basically, with the grouping of labels determined by the subtrees, and the ordering of these groups determined by subtree order as described before.